src/v/serde/envelope.h

```diff
@@ -38,12 +39,36 @@ struct compat_version {
  */
 template<typename T, typename Version, typename CompatVersion>
 struct envelope {
-    bool operator==(envelope const&) const = default;
-    auto operator<=>(envelope const&) const = default;
     using value_t = T;
     static constexpr auto redpanda_serde_version = Version::v;
     static constexpr auto redpanda_serde_compat_version = CompatVersion::v;
     static constexpr auto redpanda_inherits_from_envelope = true;
+
+    // The operator definitions below may look strange, but they have the role
+    // of ensuring that serde types (which inherit from envelope) do not use
+    // the envelope comparison operators.
+    //
+    // The implementation is templated in order to distinguish the implicit call
+    // when the operator is generated for the base class (will have const
+    // serde::envelope& on both sides) and accidental calls from sub classes
+    // that do not override the operator (will have the derived template type on
+    // both sides).
+
+    template<typename O>
+    requires(std::is_same_v<
+             std::remove_cvref_t<O>,
+             envelope<T, Version, CompatVersion>>)
+    bool operator==(O const&) const {
+        return true;
+    }
+
+    template<typename O>
+    requires(std::is_same_v<
+             std::remove_cvref_t<O>,
+             envelope<T, Version, CompatVersion>>)
+    std::strong_ordering operator<=>(O const&) const {
+        return std::strong_ordering::equal;
+    }
```

Comment:
https://gcc.godbolt.org/z/dzPa3b4jo maybe this works also here?

Reply:
Maybe I didn't do a great job of explaining. The issues is that this (https://gcc.godbolt.org/z/3drErKWzc) compiles.

Reply:
sorry i got the last epiphany after your message https://gcc.godbolt.org/z/jdha7zMT1

Reply:
I tried it out, but it doesn't work for weird stuff like https://gcc.godbolt.org/z/54cGbq9he, which we do in the serde tests.

Reply:
ah, i see. tried something else and it didn't work.
going with the template route, i can just suggest std::same_as to be a bit more concise https://gcc.godbolt.org/z/er1Esdvhj

Reply:
@andijcr - interesting idea with the `friend T` trick. I tried to get it to work with `private:` but didn't consider this approach which seems like a good one to keep in mind for this problem in the specific case of CRTP (though it would open up all the private members of the base to the derived case which might be undesirable in the general case though not here as everything else is public anyway).